key.go

```diff
@@ -272,6 +274,31 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
 	return ioutil.ReadAll(asAnyBio(bio))
 }
 
+// EngineLoadPrivateKey loads a private key by id
+// the id is a pkcs#11 URI https://tools.ietf.org/html/rfc7512#section-2.3
+// Engine comes from e.g.: e,err:=openssl.EngineById("pkcs11")
+func EngineLoadPrivateKey(e *Engine, id string) (PrivateKey, error) {
+	if e == nil {
+		return nil, errors.New("ENGINE_load_private_key cannot be called with NULL engine")
+	}
+
+	keyID := C.CString(id)
+	defer C.free(unsafe.Pointer(keyID))
+
+	key := C.ENGINE_load_private_key(e.e, keyID, nil, nil)
+	if key == nil {
+		return nil, errors.New("cannot load private key, ENGINE_load_private_key error")
+	}
+
+	// engine_ref trick inspired by the work of Renato Aguiar https://github.com/renatoaguiar
+	// it prevents the engine to be freed while we still use the key.
+	p := &pKey{key: key, engine_ref: e}
+	runtime.SetFinalizer(p, func(p *pKey) {
```

Comment:
On a second thought; this is VERY dangerous. Looking at the other functions with `pKey` receiver, there's no call to `runtime.KeepAlive` (or `runtime.LockOSThread`) after last reference to this struct, potentially making the garbage collector routine execute the finalizer before the last cgo call returns (which might lead to undefined behavior).
Consider the following example:
```
package main

/*
#include <stdio.h>
#include <unistd.h>

void stuff(char *p) // sleep for a sec and print string
{
	sleep(1); write(1, p, 4);
}
*/
import "C"

import (
	"runtime"
	"time"
	"unsafe"
)

type Test struct {
	Foo *C.char
}

func (t *Test) DoTheStuff() {
	C.write(1, unsafe.Pointer(t.Foo), 4)
	C.stuff(t.Foo) // last reference here!
}

// A terrible example of the dangers of runtime.SetFinalizer
func main() {
	str := []byte("foo\n")
	t := &Test{Foo: (*C.char)(unsafe.Pointer(&str[0]))}
	runtime.SetFinalizer(t, func(t *Test) { str[0] = 'p' }) // imagine C.free()
	go func() {
		for {
			runtime.GC()
			time.Sleep(time.Millisecond * 500)
		}
	}()
	t.DoTheStuff()
}
```
When executed, the finalizer will run while the cgo call is in progress.

Reply:
discussed on slack. thanks @alfrunes